docs/source/containers.rst

```diff
@@ -1,34 +1,303 @@
+default of ``None`` with the same name as the class attribute. A quick way
+to see how this information will be put into the the database is taking a look
+at ``dict(item)``:
+
+.. ipython:: python
+
+    item = MyItem(name="my_item", model_no="QABC1234")
+    dict(item)
+
+As shown in the example above, using the EntryInfo keywords, you can put a
+short doc string to give a better explanation of the
+field, and also enforce that user enter a specific format of information.
+
+While the user will always be able to enter ``None`` for the attribute, if a
+real value is given it will be checked against the specified ``enforce``
+keyword, raising ``ValueError`` if invalid. Here is a table for how the
+:class:`.EntryInfo` check the type
+
+=======   ===========================
+Enforce   Method of enforcement
+=======   ===========================
+None      Any value will work
+type      type(value)
+list      list.index(value)
+regex     regex.match(value) != None
+=======   ===========================
+
+Fields that are important to the item can be marked as mandatory
+(``optional=False``); these should have no default value.
+
+When entering information you will not neccesarily see a difference between
+optional and mandatory :class:`.EntryInfo`, however the database client will
+reject the item if these fields do not have the requisite values set.
+
+Loading your Object
+^^^^^^^^^^^^^^^^^^^
+
+A container's primary role is containing the information necessary to load the
+Python representation of an object.
+
+Internally, happi keeps track of all containers by way of its registry, the
+:class:`~happi.containers.HappiRegistry`.
+
+This information is stored as a ``device_class``, ``args`` and ``kwargs``. The
+former stores a string that indicates the Python class of the item, the other
+two indicate the information that is needed to instantiate it. With this
+information both :func:`.from_container` and :func:`.load_device` will handle
+importing modules and instantiating your object.
+
+.. note::
+
+    happi will attach the original metadata with a fixed attribute name ``.md``
+    to your object.  You can use this to keep track of the container metadata
+    used to instantiate your object.
+
+    This can be disabled by setting ``attach_md`` to ``False`` in
+    :func:`.from_container`.
+
+Often information contained in the ``args`` or ``kwargs`` will be duplicated in
+other parts of the container. For instance most ``ophyd`` objects will want a
+``name`` and ``prefix`` on initialization. Instead of repeating that
+information you can just use a template and have the information automatically
+populated for you by the container itself. For instance, in the aforementioned
+example ``container.args = ["{{name}}"]`` would substitute the attribute
+listed as ``container.name`` in as an argument. If the template contains the
+substituted attribute alone the type will also be converted.
```

Comment:
```suggestion
Often, information contained in the ``args`` or ``kwargs`` will be duplicated
in other parts of the container. For instance, most ``ophyd`` objects will want
a ``name`` and ``prefix`` on initialization. Instead of repeating that
information you can just use a template and have the information automatically
populated for you by the container itself. For instance, in the aforementioned
example, ``container.args = ["{{name}}"]`` would substitute the value of
``container.name`` in as an argument. If the template contains the substituted
attribute alone, the type will also be converted.
```